lib/core/subrequest.c

```diff
@@ -0,0 +1,128 @@
+        if (subreq->super.parent->conn->callbacks->log_.category.name != NULL) {                                                   \
+            return subreq->super.parent->conn->callbacks->log_.category.name(subreq->super.parent);                                \
+        } else {                                                                                                                   \
+            return h2o_iovec_init(NULL, 0);                                                                                        \
+        }                                                                                                                          \
+    }
+
+DEFINE_LOGGER(ssl, protocol_version)
+DEFINE_LOGGER(ssl, session_reused)
+DEFINE_LOGGER(ssl, cipher)
+DEFINE_LOGGER(ssl, cipher_bits)
+DEFINE_LOGGER(ssl, session_id)
+DEFINE_LOGGER(http1, request_index)
+DEFINE_LOGGER(http2, stream_id)
+DEFINE_LOGGER(http2, priority_received)
+DEFINE_LOGGER(http2, priority_received_exclusive)
+DEFINE_LOGGER(http2, priority_received_parent)
+DEFINE_LOGGER(http2, priority_received_weight)
+DEFINE_LOGGER(http2, priority_actual)
+DEFINE_LOGGER(http2, priority_actual_parent)
+DEFINE_LOGGER(http2, priority_actual_weight)
+
+#undef DEFINE_LOGGER
+
+h2o_subreq_t *h2o_subrequest_create(h2o_req_t *parent, size_t sz)
+{
+    static const h2o_conn_callbacks_t callbacks = {
+        get_sockname,    /* stringify address */
+        get_peername,    /* ditto */
+        NULL,            /* push */
+        get_socket,      /* get underlying socket */
+        get_debug_state, /* get debug state */
+        {{
+            {log_protocol_version, log_session_reused, log_cipher, log_cipher_bits, log_session_id},                       /* ssl */
+            {log_request_index},                                                                                           /* http1 */
+            {log_stream_id, log_priority_received, log_priority_received_exclusive, log_priority_received_parent,
+                log_priority_received_weight, log_priority_actual, log_priority_actual_parent, log_priority_actual_weight} /* http2 */
+        }}};
+
+    struct st_h2o_subreq_t *subreq = h2o_mem_alloc(sz);
+    subreq->conn.ctx = parent->conn->ctx;
+    subreq->conn.hosts = parent->conn->hosts;
+    subreq->conn.connected_at = parent->conn->connected_at;
+    subreq->conn.id = parent->conn->id;
+    subreq->conn.callbacks = &callbacks;
+    h2o_init_request(&subreq->super, &subreq->conn, NULL);
+
+    subreq->super.parent = parent;
+    subreq->super.input = parent->input;
+    subreq->super.hostconf = parent->hostconf;
+    subreq->super.pathconf = parent->pathconf;
+    subreq->super.version = parent->version;
+    subreq->super.error_logs = parent->error_logs;
+    subreq->super.num_delegated = parent->num_delegated;
+    subreq->super.num_reprocessed = parent->num_reprocessed;
+    subreq->super.processed_at = parent->processed_at;
+
+    return subreq;
+}
```

Comment:
Assuming that this is the only function to be defined in the source file (see my comments below about other functions), I think we can move the definition to request.c and remove subrequest.c entirely.